### DISCUSSION TITLE
getResidualCov

### DISCUSSION THREAD

**Discussion by AlainZuur** [JenniNiku/gllvm#135] [General]
Hello,

The example here:

https://jenniniku.github.io/gllvm/articles/vignette1.html

Just above the heading "Incorporating functional traits into fourth corner models", you do this:

rcov <- getResidualCov(fit_env, adjust = 0)
rcov0 <- getResidualCov(fit1lv, adjust = 0)
rcov0$trace; rcov$trace
1 - rcov$trace / rcov0$trace

Why is the adjust = 0, and not 1? It is a NB model.

I can vaguely see why you would do that, but I can't bring it under words. As a side question....is it possible to set the NB variance parameters theta_j to a pre-set value? When you do ordinary negative binomial GLM, the variance parameter theta is set to the value of the full model when doing things like drop1. Would it not make sense to set the theta_j values of sub-models to the theta_j of the full model? I believe glmmTMB can do that.

And that ratio business of these two traces to calculate the explained variance of the covariates......is that also valid if there is a random effect for rows in the model?

Alain

**Comment by BertvanderVeen** [comment]
Hello Alain, apologies for the slow response.

gllvm does not explicitly parameterize the NB in terms of its variance, but either how it is not currently possible to fix the dispersion parameters to a predetermined value. It is possible to collect them across all species (i.e., estimate only one value) or group-specific (e.g., estimate dispersion per functional groups).

If there are random effects in the model, and you are interested in the variance explained only by the unconstrained latent variables, there is no need to account for the variance of the random row effects. In the grand scheme of the whole model, that is different. Let's say we have the model:

$\eta_{ij}  = \beta_{0j}+ \boldsymbol{z}_i^\top\boldsymbol{\gamma}_j, \qquad \text{where } \boldsymbol{z}_I \sim \mathcal{N}(\textbf{0}, \textbf{I})$

where the last term represent the effects due to the LVs and random row effect (for the latter $\gamma_j = \sigma$, i.e., its standard deviation). So that we have:

$\text{var}(\boldsymbol{z}_i^\top\boldsymbol{\gamma}_j,\boldsymbol{z}_i^\top\boldsymbol{\gamma}_j) = \boldsymbol{\gamma}_j\boldsymbol{\gamma}_j \qquad \forall j$

The trace is a sum over the species $j$ so that for $q = \ldots d+1$ with $d$ number of latent variables this results in the expression:

$\sum\limits^p\limits_{j=1} \sum\limits^{d+1}\limits_{q=1} \gamma_{qj}^2 = p\sigma^2+\sum\limits^p\limits_{j=1} \sum\limits^d\limits_{q=1} \gamma_{qj}^2$.

Similarly with the `adjust = 0/1/2` argument. If you are interested in "just" examining the covariation due to the latent effects alone, `adjust = 0` is probably most appropriate. In the grand scheme of the model, `adjust = 1/2` is potentially a better choice.

There should be no issue following the same ideas as in the vignette when including covariates, as long as all sources of residual variation are included in the calculation.

Hope that helps.
Bert
